### DISCUSSION TITLE
Frequent disconnect when using nginx reverse proxy

### DISCUSSION THREAD

**Discussion by billtran1632001** [Comfy-Org/ComfyUI#2786] [Q&A] [answered]
I set up nginx reverse proxy using [linuxserver.io SWAG docker image](https://docs.linuxserver.io/general/swag/) for ComfyUI with basic authentication. But it kept disconnecting and reconnecting very frequently, about every few minutes or so. I haven't timed it yet. First glance at Firefox network debug shown that the reconnecting popup at the same time a websocket request was sent. I don't have any experience or knowledge in this subject, so does anybody have any idea?

The config snippet for ComfyUI is as follows:

```nginx
server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;

    server_name comfyui.*;

    auth_basic           "ComfyUI";
    auth_basic_user_file /config/nginx/proxy-confs/conf/auth;

    include /config/nginx/ssl.conf;

    client_max_body_size 0;

    location / {
        include /config/nginx/proxy.conf;
        include /config/nginx/resolver.conf;

        set $upstream_app 192.168.100.20;
        set $upstream_port 8188;
        set $upstream_proto http;

        proxy_pass $upstream_proto://$upstream_app:$upstream_port;
        proxy_set_header Host $upstream_app:$upstream_port;
        proxy_set_header X-Forwarded-For $host;
    }

    location /ws {
        # include /config/nginx/proxy.conf;
        # include /config/nginx/resolver.conf;

        set $upstream_app 192.168.100.20;
        set $upstream_port 8188;
        set $upstream_proto http;

        proxy_pass $upstream_proto://$upstream_app:$upstream_port;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_set_header Host $upstream_app:$upstream_port;
        proxy_set_header X-Forwarded-Host $host;

        proxy_read_timeout  36000s;
    }
}
```

**Comment by hydrusbeta** [comment]
Does the disconnect happen immediately after the call to /ws or a little bit after?
~~I see you set a proxy_read_timeout of 10 hours, but did not set values for proxy_connect_timeout or proxy_send_timeout. The defaults for those are both 60 seconds, so I'm wondering if one of those might be timing out.~~
Edit: I don't think the issue is with proxy_connect_timeout or proxy_send_timeout because I was able to get an Nginx reverse proxy working with a fairly similar configuration:
```nginx
server {
  listen 8189 default_server;
  server_name _;
  location / {
     proxy_pass http://comfyui:8188;
     proxy_set_header Host $host;
  }
  location /ws {
     proxy_pass http://comfyui:8188;
     proxy_http_version 1.1;
     proxy_set_header Host $host;
     proxy_set_header Connection 'Upgrade';
     proxy_set_header Upgrade $http_upgrade;
     proxy_read_timeout 36000s;
}
```
So I have two new questions: 
1. Does the disconnect happen predictably or is it random?
2. What's in /config/nginx/proxy.conf and /config/nginx/resolver.conf? Maybe something is going on in one of those two files that's messing with the websocket connection.

**Reply by hydrusbeta** [reply]
Looking into this some more, I downloaded the linuxserver/swag docker image and looked at proxy.conf and resolver.conf. They contain a bunch of default settings. I tried applying them all to my configuration and everything worked just fine, so I don't think those files are a problem. 
I'm not sure what could be causing a disconnect in your case. Maybe it has something to do with SSL? Are you able to test with an insecure connection over port 80? 
Other than that, you could try adding proxy_send_timeout and proxy_connect_timeout values though they didn't cause any issue for me.

**Reply by billtran1632001** [reply, marked as answer]
After some testing, it seems that it was a VPN fault. I had Cloudflare WARP running on the client machine (to unblock some medium.com sites). Without WARP activated, I don't see the disconnection anymore.

**Reply by hydrusbeta** [reply]
Thanks for following up. Glad to hear you have resolved the issue!

**Comment by pjbrunet** [comment]
I found that you don't need a separate location block for /ws 

The key line here for me was `proxy_set_header Upgrade $http_upgrade;` otherwise wss: requests were blocked. 

I'm not sure if all of these lines are necessary, but this is working for me :-) 
```

location / {
	proxy_pass http://127.0.0.1:8188/;
	proxy_http_version 1.1;
	proxy_set_header Host $host;
	proxy_set_header Connection "Upgrade";
	proxy_set_header Upgrade $http_upgrade;
	proxy_read_timeout  36000s;
}
```

**Comment by ipaqmaster** [comment]
After some trial and error with nginx tweaking and saved workflows not loading on click - 

This location block seems to be working for me. With no other location blocks. Just this one alone at the bottom of my tls server directive.

```
  location / {
    proxy_pass http://127.0.0.1:8188$request_uri;
    proxy_set_header Host $host;
    proxy_http_version 1.1;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Sec-WebSocket-Extensions $http_sec_websocket_extensions;
    proxy_set_header Sec-WebSocket-Key $http_sec_websocket_key;
    proxy_set_header Sec-WebSocket-Version $http_sec_websocket_version;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "Upgrade";
  }
```

I'm posting this because this is the top google result for nginx reverse proxy configurations for ComfyUI and some of the included configs in this thread also had problems when loading workflows and such.
